File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0547.py
class Solution:
    #-----------------------------------------------------
    # Simply do BFS & find connected components
    #
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)

        BLACK = 0
        WHITE = 1
        color = [WHITE]*n
        
        noProvinces = 0
        for i in range(n):
            if (color[i] == BLACK): continue

            noProvinces += 1
            color[i] = BLACK
            Q = []
            Q.append(i);
            while (len(Q) > 0):
                u = Q.pop(0)
                
                for v in range(len(isConnected[i])):
                    if (isConnected[u][v] and color[v] == WHITE):
                        color[v] = BLACK
                        Q.append(v)
        
        return noProvinces

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/17-Graphs-BFS/LeetCode/LeetCode-0547.cpp
class Solution {
public:
    ///-----------------------------------------------------
    /// Simply do BFS & find connected components
    ///
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();

        #define BLACK 0
        #define WHITE 1
        vector<char> color(n, WHITE);
        
        int numProvinces = 0;
        for (int i=0; i<n; i++){
            if (color[i] == BLACK) continue;

            numProvinces++;
            color[i] = BLACK;
            queue<int> Q;
            Q.push(i);
            while (!Q.empty()){
                int u = Q.front(); Q.pop();
                
                for (int v=0; v<isConnected[i].size(); v++){
                    if (isConnected[u][v] && color[v] == WHITE){
                        color[v] = BLACK;
                        Q.push(v);
                    } // end-if
                } // end-for
            } // end-while
        } // end-for
        
        return numProvinces;
    } // end-findCircleNum
};